spec/21-http_header_format_rationale.md

```diff
@@ -16,6 +16,103 @@ Making `trace-flags` optional doesn't save a lot, but makes specification more c
+   interoperability.
+3. There must be a way forward to implement Trace Context protocol by systems
+   that doesn't support long trace identifiers.
+
+### How 64bit systems may switch to Trace Context
+
+#### Absolute minimum
+
+The absolute minimum requirement from specification perspective is to receive
+and send a valid `traceparent` header. Systems operating with shorter `trace-id`
+may use only a subset of `trace-id` bytes to read and set `trace-id`. This
+behavior will break interoperability with vendors and platforms operating with
+the longer identifiers. If only a subset of `trace-id` bytes were read on
+incoming requests and sent with the outgoing call, for the systems operating
+with longer identifiers incoming and outgoing `trace-id` will not match. These
+systems will identify this situation as a restarted trace.
+
+Note, that vendors and platforms may implement a special logic to interoperate
+with the systems like this.
+
+Specification uses "SHOULD language" when asking to fill up extra bytes with
+random numbers instead of zeros. Typically, tracing systems that will not
+propagate extra bytes of incoming `trace-id` will not follow this ask and will
+fill up extra bytes with zeros.
+
+#### Linking to the longer thread-id
+
+As a minor step to improve interoperability between tracing systems, system that
+operates with shorter identifiers may record a longer incoming `trace-id` as a
+property of a telemetry item representing the incoming request.
+
+#### Propagating extra bytes of trace-id
+
+Preserving the `trace-id` unchanged is a major improvement in interoperability
+of tracing systems using different number of `trace-id` bytes. It is typical
+that tracing systems may propagate extra information from incoming request to
+outgoing calls. While recording of these extra bytes to the tracing systems
+backend is not possible.
+
+If tracing system can propagate these extra bytes, than it MUST do it.
+
+When tracing system has this capability, specification suggests that extra bytes
+are fill out with random numbers instead of zeros. This requirement helps to
+validate that tracing systems implements propagation from incoming request to
+outgoing calls correctly.
+
+#### Propagating tracestate as well
+
+Note, that even greater interoperability will be achieved with propagating a
+`tracestate` header. Dropping this header may break vendor specific distributed
+tracing scenarios. But this behavior conforms to the specification and older
+tracing systems may do it.
+
+As noted in previously, these tracing systems must do a best effort of
+propagating this header even if it will not be recorded to the tracing system
+backend.
+
+## Trace ID randomization "left padding"
+
+Specification explains the "left padding" requirement for the random `trace-id`
+generation. Tracing systems will implement various algorithms that use
+`trace-id` as a base to make a sampling decision. Typically, it will be some
```

Comment:
Ahhhh ... is this the main reason for recommending random 64 bits for backfill? I can understand how this would be beneficial for such sampling algorithms, but I would suggest that interoperability-by-default with existing brown field systems is a *much* better tradeoff.

(Also, how is "typically" determined here? How common is this really? How many major large-userbase tracers have this functionality? How many users actually use it? I'd be surprised if the userbase using this kind of sampling algorithm is (1) really big enough to warrant breaking default interoperability with 64 bit systems, and (2) would mostly use the leftmost bytes such that they would actually be negatively affected by backfilling with zeros.)

Reply:
It seems it's the only reason, and it doesn't hold water for me. Someone mentioned elsewhere that some systems may use UUID algorithms for generating trace IDs, and the left-most bytes could be composed of a timestamp - not random at all.

If systems want to avoid RNG by reusing randomness in the trace ID, an easier way to do that is XOR the left and right 8 bytes of the full 16-bytes ID and stop caring about which side of it contains more randomness.

Reply:
> Ahhhh ... is this the main reason for recommending 

we need to keep this rationale file up to date. It really helps I think.

Reply:
As for the recommendation - we had this issue in Microsoft with the "bad ordering" of randomness and I saw a few implementations which only looking at 4 bytes to calculate sampling priority. However I personally don't feel strongly about this clause. Exactly because of these reasons that it is hard to enforce randomness.